diff -Naur proxy/common/scripts/run.sh proxy-new/common/scripts/run.sh
--- proxy/common/scripts/run.sh	2023-04-08 21:12:05.896147208 +0800
+++ proxy-new/common/scripts/run.sh	2023-04-08 20:33:51.935437889 +0800
@@ -40,6 +40,7 @@
 MOUNT_ENVOY_SOURCE="${MOUNT_ENVOY_SOURCE:-`cd $MOUNT_SOURCE/../envoy;pwd`}"
 MOUNT_PACKAGE_SOURCE="${MOUNT_PACKAGE_SOURCE:-`cd $MOUNT_SOURCE/../package;pwd`}"
 MOUNT_ROOT_SOURCE="${MOUNT_ROOT_SOURCE:-`cd $MOUNT_SOURCE/..;pwd`}"
+MOUNT_PLUGINS_SOURCE="${MOUNT_PLUGINS_SOURCE:-`cd $MOUNT_SOURCE/../../plugins/wasm-cpp;pwd`}"
 
 read -ra DOCKER_RUN_OPTIONS <<< "${DOCKER_RUN_OPTIONS:-}"
 
@@ -64,6 +65,7 @@
     --mount "type=bind,source=${MOUNT_SOURCE},destination=/work" \
     --mount "type=bind,source=${MOUNT_ROOT_SOURCE}/..,destination=/parent" \
     --mount "type=bind,source=${MOUNT_ENVOY_SOURCE},destination=/envoy" \
+    --mount "type=bind,source=${MOUNT_PLUGINS_SOURCE},destination=/wasm-cpp" \
     --mount "type=volume,source=go,destination=/go" \
     --mount "type=volume,source=gocache,destination=/gocache" \
     --mount "type=volume,source=cache,destination=/home/.cache" \
