diff --git a/source/common/http/headers.h b/source/common/http/headers.h
index a7a8a3393e..6af4a2852d 100644
--- a/source/common/http/headers.h
+++ b/source/common/http/headers.h
@@ -123,7 +123,7 @@ public:
     const LowerCaseString TriCostTime{"req-cost-time"};
     const LowerCaseString TriStartTime{"req-start-time"};
     const LowerCaseString TriRespStartTime{"resp-start-time"};
-    const LowerCaseString EnvoyOriginalHost{"original-host"};
+    const LowerCaseString EnvoyOriginalHost{"x-envoy-original-host"};
     const LowerCaseString HigressOriginalService{"x-higress-original-service"};
   } AliExtendedValues;
 #endif
